pages/blog/extracting-all-mongodb-fields.mdx

```diff
@@ -0,0 +1,122 @@
+
+Before starting to get into the solution for this, let's discuss a couple of complicators of using a flexible schema database. Since there is no restriction for adding new fields it is impossible to guarantee that all documents have the same fields without reading all of them. Also, MongoDB allows related data to be nested within a single data structure and this fact increases the complexity of each document and, consequently, its queries. So, dealing with such a type of schema organization demands different thinking of how to store and retrieve data.
+
+In our case, the information to be retrieved is all the field names including the nested ones. Ok, let’s split this problem into two parts: get all first-level fields and get all nested fields with dot notation (i.e. parent.child). For the first one, there are quite a few alternatives to reach the same results, such as:
+
+### Map-Reduce
+
+Map-reduce operations use custom JavaScript functions to map, or associate, values to a key. If a key has multiple values mapped to it, the operation reduces the values for the key to a single object. Here is an example of how to get all first level fields of a collection using Map-Reduce:
+
+```
+db.runCommand({
+ "mapreduce": "my_collection", "map": function() {
+   for (var key in this) {
+     emit(key, null);
+   }
+ }, "reduce": function(key, stuff) {
+   return null;
+ }, "out": "my_collection_keys"
+});
+```
+
+After that, run the distinct command on the resulting collection to get the a set containing all keys:
+
+```
+db.my_collection_keys.distinct("_id")
+```
+### Mongo shell
+Alternatively, you can get the same result from a mongo shell client:
+```
+var allKeys = {};
+db.my_collection.find().forEach(function(doc) {
+ Object.keys(doc).forEach(function(key) {
+   allKeys[key] = 1;
+ });
+});
+```
+
+### Aggregation Pipeline
+Also, you can use the aggregation pipeline which works with all drivers that support the aggregate framework. Within your aggregation pipeline, you can use the $sample and/or $limit in order to reduce the overhead for large collections.
+
+```
+const firstLevelFieldsResult = await db.my_collection.aggregate([
+ { $project: { keyValue: { $objectToArray: "$$ROOT" } } },
+ { $unwind: "$keyValue" },
+ { $group: { _id: null, keys: { $addToSet: "$keyValue.k" } } }
+]);
+```
+
+There are several other ways to achieve the same result, but the aggregation pipeline solution provides better performance (at least than the Map-reduce solution) and usability than the other ones.
+
+Great!! The first part of our problem is solved. Now, how can we get the nested fields? To solve this, we’re going to get deep into these nested fields and extract all the names one by one. Yup, you’re right, this is a very expensive task and we need to limit the number of documents to get into.
+
+```
+const docs = await db.my_collection.aggregate([
+ { $sample: { size: 10 } },
```

Comment:
What's cool about this is that this lets you control the "accuracy" of your search - add more `size` if your schema has dramatically different object types.  I imagine that most of Grouparoo's users are using something Mongose to enforce a schema... so hopefully it's not so crazy.

Reply:
Yes, I think the documents will only be slightly different.
But, if we receive any complaints regarding missing fields we can perform a single-time check (trying to determine how different the docs are) to define the sample size.